CMakeLists.txt

```diff
@@ -368,3 +369,26 @@ if (TESTS)
 	add_subdirectory(Test)
 endif (TESTS)
 
+CHECK_CXX_SOURCE_COMPILES("
+	#include <stdlib.h>
+	extern \"C\" {
+	__attribute__((weak))
+	void *__cxa_allocate_exception(size_t thrown_size) noexcept;
+	}
+	#include <exception>
+	int main() { return 0; }" CXA_ALLOCATE_EXCEPTION_NOEXCEPT_COMPILES)
+
+if (CXA_ALLOCATE_EXCEPTION_NOEXCEPT_COMPILES)
+	add_definitions(-DCXA_ALLOCATE_EXCEPTION_SPECIFIER=noexcept)
+else ()
+	add_definitions(-DCXA_ALLOCATE_EXCEPTION_SPECIFIER=)
+endif ()
+
+CHECK_CXX_SOURCE_COMPILES("
+	#include <exception>
+	namespace std { class type_info { }; }
+	int main() { return 0; }" TYPE_INFO_REDEFINITION_COMPILES)
+
+if (NOT TYPE_INFO_REDEFINITION_COMPILES)
```

Comment:
Please put this in a separate PR once you've explained what problem it is solving.